Translation of episode 6: loops and functions. (#46)
Carpentries-ES/python-ecology-lesson-es: _episodes/06-loops-and-functions.md

~~~
{: .output}

El comando `os.listdir` es equivalente a usar `ls` en la terminal.

En episodios anteriores, vimos cómo usar la librería pandas para cargar la
información sobre las especies en la memoría como un **DataFrame**, vimos cómo
seleccionar un subconjunto de esos datos usando ciertos criterios, y vimos
también cómo escribir esa información en un archivo CSV. Escribamos un
**script** que realice esos tres pasos en secuencia para el año 2002:

~~~
import pandas as pd

# Cargamos los datos en un DataFrame
surveys_df = pd.read_csv('data/surveys.csv')

# Seleccionamos solo los datos del año 2002
surveys2002 = surveys_df[surveys_df.year == 2002]

# Escribimos el nuevo DataFrame en un archivo CSV
surveys2002.to_csv('data/yearly_files/surveys2002.csv')
~~~
{: .language-python}

Para crear los archivos con los datos anuales, podemos repetir estos últimos dos
comandos una y otra vez, una por cada año de información. Sin embargo, repetir
código no es ni elegante ni práctico, y hace muy probable que introduzcamos
errores en nuestro código. Queremos convertir lo que acabamos de escribir en un
bucle que repita estos últimos dos comandos para cada año en nuestro dataset.

> 

Author: Oh, mi error. Lo corregiré después de revisar los cambios.

~~~
{: .language-python}

Mira dentro del directorio `yearly_files` y verifica un par de los archivos que
acabaste de crear para confirmar que todo funcionó como esperabas.

## Escribiendo nombres de archivo únicos

Nota que en el código anterior creamos un nombre de archivo único para cada año.

~~~
filename = 'data/yearly_files/surveys' + str(year) + '.csv'
~~~
{: .language-python}

Entendamos las partes de este nombre:

* La primera parte es simplemente un texto que especifica el directorio en el que
  vamos a guardar nuestro archivo (data/yearly_files/) y la primera parte del
  nombre del archivo (surveys): `'data/yearly_files/surveys'`
* Podemos concatenar esto con el valor de una variable, en este caso `year`, al
  usar el signo más y la variable que le queremos añadir al nombre del archivo:
  `+ str(year)`
* Por último añadimos la extensión del archivo usando otra cadena de caracteres:
  `+ '.csv'`

Nota que usamos comillas sencillas para añadir las cadenas de caracteres y que la
variable no está entre comillas. Este código produce la cadena de caracteres
`data/yearly_files/surveys2002.csv` que contiene tanto el **path** como el
nombre del archivo.

> > ## Challenge - Modificando bucles

Reviewer: ```suggestion
> ## Desafío - Modificando bucles
```

>

> > 1. Algunos de los surveys que guardaste tienen datos faltantes (tienen valores

Reviewer: ```suggestion
> 1. Algunas de las encuestas que guardaste tienen datos faltantes (tienen valores
```

>    nulos que salen como `NaN` - No es un número (en inglés) - en los DataFrames

> >    y no salen en los archivos de texto). Modifica el bucle for para que las

Reviewer: ```suggestion
>    y no salen en los archivos de texto). Modifica el bucle **for** para que las
```

> >    entradas que tengan valores nulos no sean incluidos en los archivos anuales.

Reviewer: ```suggestion
>    entradas que tengan valores nulos no sean incluidas en los archivos anuales.
```

>
> 2. Supongamos que solo quieres revisar los datos cada cierto múltiplo de años.
>    ¿Cómo modificarías el bucle para generar un archivo de datos cada cinco
>    años comenzando desde 1977?
>
> 3. En vez de separar la información por años, un colega tuyo quiere hacer el
>    análisis separando por especies. ¿Cómo escribirías un único archivo CSV
>    por cada especie?
{: .challenge}

> ## Construyendo código modular y reusable usando funciones

Reviewer: ```suggestion
## Construyendo código modular y reusable mediante el uso de funciones
```

Author: Hola, Fer. No sé qué opines, pero: ¿no es mejor usar en títulos la mínima cantidad de palabras posibles?

> Supón que separar archivos enormes en archivos anuales induviduales es una tarea

Reviewer: ```suggestion
Supón que separar archivos enormes en archivos anuales individuales es una tarea
```

que tenemos que realizar frecuentemente. Podríamos escribir un bucle **for**
como el que hicimos arriba cada vez que lo necesitemos, pero esto tomaría
mucho tiempo y podría introducir errores en el código. Una solución más elegante
sería crear una herramienta reusable que realice esta tarea con el mínimo
esfuerzo del usuario. Para hacerlo, vamos a convertir el código que ya
escribimos en una función.

> Las funciones son pedazos de códigos reusables y autónomos que pueden ser

Reviewer: ```suggestion
Las funciones son piezas de código reusable y autónomo que pueden ser
```

> llamados con un solo comando. Están diseñados para aceptar argumentos como

Reviewer: ```suggestion
llamadas mediante un solo comando. Están diseñadas para aceptar argumentos como
```

Author: Wow, estoy teniendo serios problemas con el género.

entrada y retornar valores, pero no necesitan hacerlo necesariamente. Las
variables declaradas adentro de las funciones solo existen mientras la función
se está ejecutando, y si una variable adentro de una función (una variable
local) tiene el mismo nombre de otra variable en alguna parte del código, la

> variable local no sobreescribe la otra.

Reviewer: De acuerdo con la RAE [no se admite la doble "e" en sobrescrito](http://lema.rae.es/dpd/srv/search?key=sobrescribir) 

```suggestion
variable local no sobrescribe a la otra.
```

Todo método usado en Python (como por ejemplo `print`) es una función, y las
librerías que importamos (`pandas`, por ejemplo) son una colección de funciones.
Solo usaremos las funciones que están contenidas en el mismo código que las
usan, pero es sencillo también escribir funciones que puedan ser usadas por
programas diferentes.

Las funciones se declaran usando la siguiente estructura general:

~~~
def this_is_the_function_name(input_argument1, input_argument2):

    # El cuerpo de la función está indentado
    # Esta función imprime los dos argumentos en pantalla
    print('The function arguments are:', input_argument1, input_argument2, '(this is done inside the function!)')

    # And returns their product
    return input_argument1 * input_argument2
~~~
{: .language-python}

La declaración de la función comienza con la palabra clave `def`, seguida del
nombre de la función y los argumentos entre paréntesis, y termina con un dos

> puntos. El cuerpo de la función está indentada justo como ocurría con los

Reviewer: ```suggestion
puntos. El cuerpo de la función está indentado justo como ocurría con los
```

~~~
{: .output}

> > ## Challenge - Funciones

Reviewer: ```suggestion
> ## Desafío - Funciones
```

>
> 1. Cambia los valores de los argumentos en la función y mira su salida.
> 2. Intenta llamar a la función usando la cantidad equivocada de argumentos
>    (es decir, diferente de 2) o sin asignar la llamada de la función a una
>    variable (sin poner `product_of_inputs =`).
> 3. Declara una variable dentro de una función y prueba a encontrar en dónde
>    existe (Pista: ¿puedes imprimirla desde fuera de la función?)
> 4. Explora qué sucede cuando una variable tiene el mismo nombre adentro y

> >    afuera de la función. ¿Qué le ocurre a la variable globar cuando cambias el

Reviewer: ```suggestion
>    afuera de la función. ¿Qué le ocurre a la variable global cuando cambias el
```

>    valor de la variable local?
{: .challenge}

Ahora podemos convertir el código para guardar archivos con datos anuales en una

> función. Hay muchas "partes" de este códugo que podemos convertir en funciones,

Reviewer: ```suggestion
función. Hay muchas "partes" de este código que podemos convertir en funciones,
```

~~~
{: .language-python}

Cambiamos el nombre del archivo CSV para diferenciarlo del que escribimos
anteriormente. Busca en el directorio `yearly_files` el archivo que creamos.
¿Hizo la función lo que esperabas que hiciera?

Sin embargo, lo que nos encantaría hacer es crear archivos para múltiples años

> sin tener que pedirlos uno a uno. Escribamos otra función que reemplaza el bucle

Reviewer: ```suggestion
sin tener que pedirlos uno a uno. Escribamos otra función que reemplace el bucle
```

> for por iterar a través de la secuencia de los años y llamar repetidamente la

Reviewer: ```suggestion
**for** simplemente iterando a través de la secuencia de años y llamando repetidamente a la
```

función que acabamos de escribir, `one_year_csv_writer`:

~~~
def yearly_data_csv_writer(start_year, end_year, all_data):
    """
    Escribe archivos CSV separados para cada año de datos.

    start_year --- el primer año de datos que queremos
    end_year --- el último año de datos que queremos
    all_data --- DataFrame con datos de varios años
    """

    # "end_year" es el último año que queremos extraer, entonces iteramos hasta end_year+1
    for year in range(start_year, end_year+1):
        one_year_csv_writer(year, all_data)
~~~
{: .language-python}

Como la gente esperará naturalmente que el año final (`end_year`) sea el último,

> el bucle for adentro de la función termina en `end_year + 1`. Al escribir el

Reviewer: ```suggestion
el bucle **for** adentro de la función termina en `end_year + 1`. Al escribir el
```

bucle entero en la función, hemos hecho una herramienta reusable para cuando
necesitemos partir un archivo de datos grande en archivos anuales. Como podemos

> especificar el primer y el último año de los que queremos crear archivos,

Reviewer: ```suggestion
especificar el primer y el último año para los cuales queremos crear archivos,
```

podemos inclusive usar esta función para crear archivos para un subconjunto de
los años disponibles. Así llamaríamos la función:

~~~
# Cargamos los datos en un DataFrame
surveys_df = pd.read_csv('data/surveys.csv')

# Creamos los archivos CSV
yearly_data_csv_writer(1977, 2002, surveys_df)
~~~
{: .language-python}

¡TEN CUIDADO! Si estás usando Jupyter Notebooks y estás modicando la función,
DEBES volver a ejecutar la celda para que la función cambiada esté disponible
para el resto del código. Nada cambiará visualmente cuando hagas esto, porque
definir una función sin *ejecutarla* no produce ninguna salida. Toda otra celda

> que usa la función (ahora cambiada) también tendrán que ser re-ejecutadas para

Reviewer: ```suggestion
que use la función (ahora cambiada) también tendrá que ser re-ejecutada para
```

cambiar su salida.

> ## Challenge- Más funciones
>
> 1. Añade dos argumentos a las funciones que escribimos que tomen el **path**
>    del directorio donde los archivos serán escritos y el root del nombre del
>    archivo. Crea un nuevo conjunto de archivos con un nombre diferente en un
>    directorio diferente.
> 2. ¿Cómo podrías usar la función `yearly_data_csv_writer` para crear un archivo
>    CSV para solo un año? (Pista: piensa sobre la sintaxis para `range`)

> > 3. Has que las funciones retornen una lista de los archivos que escribieron.

Reviewer: ```suggestion
> 3. Haz que las funciones retornen una lista de los archivos que escribieron.
```

>    Hay muchas formas en las que puedes hacer esto (¡y deberías intentarlas
>    todas!): cualquiera de las dos funciones podría imprimir algo en pantalla,
>    cualquiera podría usar `return` para retornar números o cadenas de
>    caracteres cada vez que se llaman, o podrías hacer una combinación de estas

> >    dos estrategias. Podrías también intentar usar la librería `os` para enlistar

Reviewer: ```suggestion
>    dos estrategias. Podrías también intentar usar la librería `os` para listar
```

~~~
{: .output}

Los "\t" en `print` son tabulaciones, y son usadas para alinear el texto y
facilitar la lectura.

> Pero ¿qué sucede si nuestro dataset no comienza en 1977 ni termina en 2002?

Reviewer: ```suggestion
Pero ¿qué sucede si nuestro **dataset** no comienza en 1977 ni termina en 2002?
```

~~~
{: .output}

Ahora los valores por defecto de los argumentos `start_year` y `end_year` en la
función `yearly_data_arg_test` son `None`. Esta es una constante incorporada en
Python que indica la ausencia de un valor - esencialmente indica que la variable

> existe en el directorio de variables (el **namespace**) de la función pero que

Reviewer: ```suggestion
existe en el directorio de nombres de variables (el **namespace**) de la función pero que
```

no corresponde a ningún objeto existente.

> ## Challenge - Variables
>
> 1. ¿Qué tipo de objeto corresponde a una variable declarada como `None`?
>    (Pista: crea una variable con el valor `None` y usa la función `type()`)
>
> 2. Compara el comportamiento de la función `yearly_data_arg_test` cuando los
>    argumentos tienen `None` como valor por defecto y cuando no tienen valores
>    por defecto.
>

> > 3. ¿Qué ocurre si solo incluimos un valor para `start_year` al llamar la función?,

Reviewer: ```suggestion
> 3. ¿Qué ocurre si solo incluimos un valor para `start_year` al llamar a la función?,
```

>    ¿puedes escribir una llamada a la función con solo un valor para `end_year`?
>    (Pista: piensa en cómo la función debe estar asignándole valores a cada uno
>    de sus argumentos - ¡esto está relacionado con la necesidad de poner los
>    argumentos que no tienen valores por defecto antes de los que sí tienen
>    valores por defecto en la definición de la función!)
{: .challenge}

> ## Condicionales if

Reviewer: ```suggestion
## Sentencias **if**
```

Author: Sí, tuve problemas al traducir la palabra statements. Sentencias me parece una excelente opción.

El cuerpo de la función anterior ahora tiene dos condicionales **if** que
revisan los valores de `start_year` y `end_year`. Los condicionales if ejecutan
un segmento de código si una condición dada es cierta. Usualmente lucen así:

~~~
a = 5

if a<0:  # ¿es cierta esta primera condición?

    # si a ES menor que cero
    print('a is a negative number')

elif a>0:  # La primera condición no es cierta, ¿la segunda?

    # si a NO ES menor que cero y ES mayor que cero
    print('a is a positive number')

else:  # No se cumplieron las dos condiciones

    # si a NO ES menor que cero y NO ES mayor que cero
    print('a must be zero!')
~~~
{: .language-python}

lo cual retornaría:

~~~
a is a positive number
~~~
{: .output}

Cambia los valores de `a` para ver cómo funciona este código. La palabra clave

> `elif` significa "si no" (en inglés, "else if"), y todos los condicionales deben

Reviewer: ```suggestion
`elif` significa "sino" (en inglés, "else if"), y todos los condicionales deben
```

~~~
{: .output}

> > ## Challenge - Modificando funciones

Reviewer: ```suggestion
> ## Desafío - Modificando funciones
```

>
> 1. Reescribe las funciones `one_year_csv_writer` y `yearly_data_csv_writer`
>    para que tengan argumentos keyword con valores por defecto.
>
> 2. Modifica las funciones de tal forma que no creen archivos para un año si
>    éste no está en los datos y que muestre una alerta al usuario (Pista: usa
>    condicionales para esto. Si quieres un reto más, ¡usa `try`!)
>

> > 3. Este código verifica si un directorio existe y lo crea si no. Añade un poco

Reviewer: ```suggestion
> 3. Este código verifica si un directorio existe, sino lo crea. Añade un poco
```

> ~~~
> {: .language-python }
>

> > 4. El código que has escrito hasta este momento usando el bucle for está

Reviewer: ```suggestion
> 4. El código que has escrito hasta este momento usando el bucle **for** está
```

> >    bastante bien, pero no necesariamente es reproducible con datasets

Reviewer: ```suggestion
>    bastante bien, pero no necesariamente es reproducible con **datasets**
```
